servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -0,0 +1,136 @@
+                // and let the higher level retries decide what to do.
+                if (!host.isActiveAndHealthy()) {
+                    return noActiveHostsException(hosts);
+                }
+                return host.newConnection(selector, forceNewConnectionAndReserve, context);
+            default:
+                return p2c(size, hosts, getRandom(), selector, forceNewConnectionAndReserve, context);
+        }
+    }
+
+    @Nullable
+    private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random random, Predicate<C> selector,
+                          boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
+        for (int j = maxEffort; j > 0; j--) {
+            // Pick two random indexes that don't collide. Limit the range on the second index to 1 less than
+            // the max value so that if there is a collision we can safety increment. We also increment if
+            // i2 > i1 to avoid biased toward lower numbers since we limited the range by 1.
+            final int i1 = random.nextInt(size);
+            int i2 = random.nextInt(size - 1);
+            if (i2 >= i1) {
+                ++i2;
+            }
+            Host<ResolvedAddress, C> t1 = hosts.get(i1);
+            Host<ResolvedAddress, C> t2 = hosts.get(i2);
+            // Make t1 the preferred host by score to make the logic below a bit cleaner.
+            if (t1.score() < t2.score()) {
+                Host<ResolvedAddress, C> tmp = t1;
+                t1 = t2;
+                t2 = tmp;
+            }
+
+            if (!forceNewConnectionAndReserve) {
+                // First we're going to see if we can get an existing connection regardless of health status. Since t1
+                // is 'better' we'll try it first. If it doesn't have any existing connections we don't fall back to t2
+                // or else we would cause a bias toward hosts with existing connections which could ultimately drive all
+                // traffic to the first host to make a connection in the case of a multiplexed session.
+                C c = t1.pickConnection(selector, contextMap);
+                if (c != null) {
+                    return succeeded(c);
+                }
+                // We now need to consider the health status and make a new connection if either
+                // host is considered healthy.
+            }
+
+            // We either couldn't find a live connection or are being forced to make a new one. Either way we're
+            // going to make a new connection which means we need to consider health.
+            final boolean t1Healthy = t1.isActiveAndHealthy();
+            final boolean t2Healthy = t2.isActiveAndHealthy();
+            if (t1Healthy) {
+                return t1.newConnection(selector, forceNewConnectionAndReserve, contextMap);
+            } else if (t2Healthy) {
+                return t2.newConnection(selector, forceNewConnectionAndReserve, contextMap);
```

Comment:
While you clarified reasoning to move re-iteration in a separate big redesign, will it make sense for the time being to follow the following logic?
1. Try `t1.pickConnection`, if not
2. Check t1 health, if ok do `t1.newConnection` if not:
3. Try `t2.pickConnection`, if not
4. Check t2 health, if ok do `t1.newConnection` if not:
5. go to the next effort